collectors/test-results/jira-xray/src/main/java/com/capitalone/dashboard/core/json/gen/TestRunJsonGenerator.java

```diff
@@ -0,0 +1,146 @@
+    public final static String KEY_EXAMPLES="examples";
+    public final static String KEY_TESTSTEPS="steps";
+    // CUMCUMBER TESTS
+    public final static String KEY_SCENARIO="scenario";
+    public final static String KEY_SCENARIO_OUTLINE="scenarioOutline";
+
+
+    private final static RendereableItemJsonGenerator COMMENTS_GENERATOR =new RendereableItemJsonGenerator();
+    private final static TestStepJsonGenerator TEST_STEPS_GENERATOR =new TestStepJsonGenerator();
+    private final static DefectJSONGenerator DEFECTS_GENERATOR =new DefectJSONGenerator();
+    private final static EvidenceJsonGenerator EVIDENCES_GENERATOR =new EvidenceJsonGenerator();
+
+
+    public JSONObject generate(TestRun testRun) throws JSONException {
+        JSONObject ex=new JSONObject();
+        // INMUTABLE FIELDS
+        ex.put(KEY_ID,testRun.getOldVersion().getId());
+        ex.put(KEY_EXECBY,testRun.getOldVersion().getExecutedBy());
+        ex.put(KEY_STARTEDON,testRun.getOldVersion().getStartedOn());
+        ex.put(KEY_ASSIGNEE,testRun.getOldVersion().getAssignee());
+        ex.put(KEY_FINISHEDON,testRun.getOldVersion().getFinishedOn());
+        // UPDATABLE FIELDS
+        ex.put(KEY_COMMENT, COMMENTS_GENERATOR.generate(testRun.getComment()).get("raw"));
+        ex.put(KEY_STATUS,testRun.getStatus().name());
+        ex.put(KEY_DEFECTS,generateDefects(testRun));
+        ex.put(KEY_EVIDENCES,generateEvidences(testRun));
+        if(testRun.getSteps()!=null) {
+            ex.put(KEY_TESTSTEPS,generateTestSteps(testRun));
+        }
+        // EXAMPLES IS CUCUMBER
+
+
+        return ex;
+    }
+
+    private JSONArray generateEvidences(TestRun testRun) throws JSONException {
+        ArrayList<JSONObject> evidences=new ArrayList<JSONObject>();
+        for(Evidence ev: testRun.getEvidences()){
+            evidences.add(EVIDENCES_GENERATOR.generate(ev));
+        }
+        return new JSONArray(evidences);
+    }
+
+    private Object generateDefects (TestRun testRun) throws JSONException{
+        if(testRun.getVersion()!=0) {
+            return generateDefectsUpdates(testRun);
+        }
+        else {
+            return generateDefectsArray(testRun);
+        }
+    }
+
+    private JSONArray generateDefectsArray(TestRun testRun) throws JSONException {
+        ArrayList<JSONObject> defects=new ArrayList<JSONObject>();
+        for(Defect def: testRun.getDefects()){
+            defects.add(DEFECTS_GENERATOR.generate(def));
+        }
+        return new JSONArray(defects);
+    }
+
+
+    //TODO: EXTRACT LOGIC AND CLEAN CODE, IMPLEMENT COMMON CLASSES AND INTERFACES
+    @SuppressWarnings("PMD.NPathComplexity")
+    private JSONObject generateDefectsUpdates(TestRun testRun) throws JSONException {
+        ArrayList<String> removes=new ArrayList<String>();
+        ArrayList<String> adds=new ArrayList<String>();
+
+        Iterable<Defect> all;
+
+        if( testRun.getVersion()!=0 && testRun.getOldVersion().getDefects()!=null) {
```

Comment:
@kumarvarun1252  use Optional to check for null in complex objects

Reply:
Optional used